include/boost/random/detail/qrng_base.hpp

```diff
@@ -0,0 +1,254 @@
+
+#ifndef BOOST_RANDOM_DETAIL_QRNG_BASE_HPP
+#define BOOST_RANDOM_DETAIL_QRNG_BASE_HPP
+
+#include <stdexcept>
+#include <vector>
+#include <limits>
+
+#include <istream>
+#include <ostream>
+#include <sstream>
+
+#include <boost/cstdint.hpp>
+#include <boost/random/detail/operators.hpp>
+
+#include <boost/throw_exception.hpp>
+
+//!\file
+//!Describes the quasi-random number generator base class template.
+
+namespace boost {
+namespace random {
+
+namespace detail {
+
+template <typename T>
+inline bool are_close(T x, T y, T eps) // |x - y| <= eps
+{
+  if (x < y) std::swap(x, y);
+  return !(x - y > eps);
+}
+
+template<typename DerivedT, typename LatticeT, typename SizeT>
+class qrng_base
+{
+public:
+  typedef SizeT size_type;
+  typedef typename LatticeT::value_type result_type;
+
+  explicit qrng_base(std::size_t dimension)
+    // Guard against invalid dimensions before creating the lattice
+    : lattice(prevent_zero_dimension(dimension))
+    , quasi_state(dimension)
+  {
+    derived().seed();
+  }
+
+  // default copy c-tor is fine
+
+  // default assignment operator is fine
+
+  //!Returns: The dimension of of the quasi-random domain.
+  //!
+  //!Throws: nothing.
+  std::size_t dimension() const { return quasi_state.size(); }
+
+  //!Returns: Returns a successive element of an s-dimensional
+  //!(s = X::dimension()) vector at each invocation. When all elements are
+  //!exhausted, X::operator() begins anew with the starting element of a
+  //!subsequent s-dimensional vector.
+  //!
+  //!Throws: range_error.
+  result_type operator()()
+  {
+    return curr_elem != dimension() ? load_cached(): next_state();
+  }
+
+  //!Fills a range with quasi-random values.
+  template<typename Iter> void generate(Iter first, Iter last)
+  {
+    for (; first != last; ++first)
+      *first = this->operator()();
+  }
+
+  //!Effects: Advances *this state as if z consecutive
+  //!X::operator() invocations were executed.
+  //!
+  //!Throws: range_error.
+  void discard(boost::uintmax_t z)
+  {
+    const std::size_t dimension_value = dimension();
+
+    // Compiler knows how to optimize subsequent x / y and x % y
+    // statements. In fact, gcc does this even at -O1, so don't
+    // be tempted to "optimize" % via subtraction and multiplication.
+
+    boost::uintmax_t vec_n = z / dimension_value;
+    std::size_t carry = curr_elem + (z % dimension_value);
+
+    vec_n += carry / dimension_value;
+    carry  = carry % dimension_value;
+
+    // Avoid overdiscarding by branchlessly correcting the triple
+    // (D, S + 1, 0) to (D, S, D) (see equality operator)
+    const bool corr = (!carry) & static_cast<bool>(vec_n);
+
+    // Discards vec_n (with correction) consecutive s-dimensional vectors
+    discard_vector(vec_n - static_cast<boost::uintmax_t>(corr));
+    // Sets up the proper position of the element-to-read
+    // curr_elem = carry + corr*dimension_value
+    curr_elem = carry ^ (-static_cast<std::size_t>(corr) & dimension_value);
```

Comment:
MSVC warns: C4146 unary minus operator applied to an unsigned type, result still unsigned.  (This should be safe to suppress).

Reply:
Yes, it is. g++ doesn't complain here at all.